packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+
+  /// Animates the scale of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 1.0.
+  Animation<double> getScaleAnimation({@required Animation<double> parent});
+
+  /// Animates the rotation of the [FloatingActionButton].
+  /// 
+  /// The animation should both start and end with a value of 0.0 or 1.0.
+  /// 
+  /// The animation values are a fraction of a full circle, with 0.0 and 1.0
+  /// corresponding to 0 and 360 degrees, while 0.5 corresponds to 180 degrees.
+  Animation<double> getRotationAnimation({@required Animation<double> parent});
+
+  @override
+  String toString() => '$runtimeType';
+}
+
+class _ScalingFabMotionAnimator extends FabMotionAnimator {
+  const _ScalingFabMotionAnimator();
+
+  @override
+  Animation<Offset> getOffsetAnimation({Offset begin, Offset end, Animation<double> parent}) {
+    return new Tween<Offset>(begin: begin, end: end).chain(new CurveTween(curve: const Threshold(0.5))).animate(parent);
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new _MaxAnimation<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
+  }
+
+  @override
+  Animation<double> getRotationAnimation({Animation<double> parent}) {
+    // Because we only see the last half of the rotation tween, 
+    // it needs to go twice as far.
+    final Tween<double> rotationTween = new Tween<double>(
+      begin: 1.0 - _kFloatingActionButtonTurnInterval * 2, 
+      end: 1.0,
+    );
+    // This rotation will turn on the way in, but not on the way out.
+    return new _MaxAnimation<double>(
+      rotationTween.animate(parent),
+      new ReverseAnimation(new CurveTween(curve: const Threshold(0.5)).animate(parent)),
+    );
+  }
+}
+
+class _MaxAnimation<T> extends CompoundAnimation<T> {
+  _MaxAnimation(Animation<T> first, Animation<T> next): super(first: first, next: next);
+
+  @override
+  T get value => math.max(first.value, next.value);
+}
+
+/// The geometry of the [Scaffold] before it finishes laying out.
+/// 
+/// The Scaffold passes this prelayout geometry to its [FabPositioner].
+/// 
+/// For a description of the Scaffold's geometry after it has
+/// finished laying out, see the [ScaffoldGeometry].
+@immutable
+class ScaffoldPrelayoutGeometry {
+  const ScaffoldPrelayoutGeometry({this.bottomSheetSize, this.contentBottom, this.contentTop, this.fabSize, this.horizontalFabPadding, this.scaffoldSize, this.snackBarSize, this.textDirection});
+
+  /// The [Size] of the [Scaffold]'s [FloatingActionButton] (if available).
+  /// 
+  /// The Scaffold will determine the fabSize if 
+  /// `Scaffold.floatingActionButton` is not null.
+  final Size fabSize;
```

Comment:
we should expand "fab" here

Reply:
otherwise people will think we are referring to the fabulous size.

Reply:
Done